### PR TITLE
Separate logic used for asserting metrics and resource attributes

### PR DESCRIPTION
**Pull request by pellared**
## Why

1. I want to add resource attribute tests to logs and spans. **The pattern from this PR can be easily reused in other MockCollectors.** _Which cannot be said for the currently used logic (before PR)._
2. I find it is cleaner to have the metrics/logs/spans test helper logic+structures separated. It can be even moved to separate classes (or methods), but I found it an overkill at this moment.
3. A good side effect is that it allows asserting for metrics and resource attributes in the same test (the assertion methods are not interfering)

## What

Separate the structures and logic used for asserting metrics and resource attributes.

## Tests

CI

### CODE REVIEW

**Comment by pellared** [comment]
I have possibly screw something as  `IntegrationTests.HttpTests.SubmitMetrics` is failing.

File: test/IntegrationTests/Helpers/MockMetricsCollector.cs [hunk 1]
```diff
@@ -292,9 +277,36 @@ private void HandleHttpRequests(HttpListenerContext ctx)
             var metricsMessage = ExportMetricsServiceRequest.Parser.ParseFrom(ctx.Request.InputStream);
             if (metricsMessage.ResourceMetrics != null)
             {
-                foreach (var metrics in metricsMessage.ResourceMetrics)
+                foreach (var rMetrics in metricsMessage.ResourceMetrics)
```
Comment (pjanotti) [inline test/IntegrationTests/Helpers/MockMetricsCollector.cs:280]: suggestion: `resourceMetrics`- I do like the short names and abbreviations but the style on the repo (and C# in general) is longer names.

Comment (pjanotti) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs:280]: anyway if changing here needs to change `sMetrics` below

**Review by pjanotti — APPROVED** [review]
LGTM, some tiny suggestions.

File: test/IntegrationTests/Helpers/MockMetricsCollector.cs [hunk 2]
```diff
@@ -37,10 +37,15 @@ public class MockMetricsCollector : IDisposable
...
     private readonly TestHttpListener _listener;
-    private readonly BlockingCollection<global::OpenTelemetry.Proto.Metrics.V1.ResourceMetrics> _metrics = new(10); // bounded to avoid memory leak
+
     private readonly List<Expectation> _expectations = new();
     private readonly List<ResourceExpectation> _resourceExpectations = new();
 
+    private readonly BlockingCollection<List<Collected>> _metricsSnapshots = new(10); // bounded to avoid memory leak
+
+    private readonly ManualResetEvent _resourceAttributesEvent = new(false); // synchronizes access to _resourceAttributes
+    private RepeatedField<KeyValue> _resourceAttributes;
```
[the remark this answers is not in the document]
Comment (pellared) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs]: Addressed 190cf53

File: test/IntegrationTests/Helpers/MockMetricsCollector.cs [hunk 3]
```diff
@@ -108,45 +109,36 @@ public void AssertExpectations(TimeSpan? timeout = null)
...
             {
-                var resourceMetrics = _metrics.Take(cts.Token); // get the metrics snapshot
+                var metrics = _metricsSnapshots.Take(cts.Token); // get the metrics snapshot
 
                 missingExpectations = new List<Expectation>(_expectations);
                 expectationsMet = new List<Collected>();
                 additionalEntries = new List<Collected>();
 
-                foreach (var scopeMetrics in resourceMetrics.ScopeMetrics)
+                foreach (var colleted in metrics)
```
[the remark this answers is not in the document]
Comment (pellared) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs]: Addressed 190cf53

File: test/IntegrationTests/Helpers/MockMetricsCollector.cs (hunk 1 above)
Comment (pellared) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs]: Addressed 190cf53

**Review by pjanotti — APPROVED** [review]

File: test/IntegrationTests/Helpers/MockMetricsCollector.cs [hunk 4]
```diff
@@ -180,36 +172,29 @@ public void AssertResourceExpectations(TimeSpan? timeout = null)
...
+            }
+
+            AssertResourceMetrics(_resourceExpectations, _resourceAttributes);
         }
         catch (ArgumentOutOfRangeException)
         {
             // CancelAfter called with non-positive value
             FailResourceMetrics(_resourceExpectations, null);
         }
-        catch (OperationCanceledException)
```
Comment (RassK) [inline test/IntegrationTests/Helpers/MockMetricsCollector.cs:202]: Just checking, the test framework itself knows to handle OperationCanceledException?

Comment (pellared) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs:202]: Nothing should throw it as I removed the uasge of CancellationTokenSource in this place

Comment (RassK) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs:202]: `var metrics = _metricsSnapshots.Take(cts.Token);` this one?

Comment (RassK) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs:202]: nvm... It's still catching there, my bad 😊

Comment (nrcventura) [reply test/IntegrationTests/Helpers/MockMetricsCollector.cs:202]: That call is in a different method now, and that exception is still handled in that method. The method that grabs the resource attributes uses `_resourceAttributesEvent.WaitOne(timeout.Value)` which does not throw that exception according to the docs.

**Review by nrcventura — APPROVED** [review]

**Review by zacharycmontoya — APPROVED** [review]

**Review by Kielek — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]

**Pull request merged**